src/global.cc

```diff
@@ -476,7 +476,7 @@ void global_scope_t::visit_man_page() const
   int status = -1;
   wait(&status);
 #endif
-  exit(0);                      // parent
+  throw error_count(0, "");     // parent
```

Comment:
Can you provide more context on this change, please?

Reply:
Now thinking about it, it may not be necessary any more.

Originally, I intended to make it easier for people to use the WASM(+JS) build as a "library" when they will usually want to call the main function multiple times. Calling `exit` had seemed to shut down the whole runtime and the user might need to reload the runtime and FS settings all over again.

However, reading more Emscripten documentation reveals that they have a dedicated flag ([`noExitRuntime`](https://emscripten.org/docs/api_reference/module.html#Module.noExitRuntime)) for this, so it is actually fine to use `exit`. Moreover, it seems Ledger does a [`quick_close`](https://github.com/ledger/ledger/blob/0c323a569f2774431fcef2e5fe4b5c425ab572b2/src/main.cc#L226) and I doubt calling `main` repeatedly might leak memory.

Overall, I think I will undo this change and maybe add some warnings on memory leaks in the to-be-draft documentation on this.  What do you think?

Reply:
This will be very interesting as the ledger python module does not work well with long running sessions as reading new data in the same runtime leads to assertion failure, because previously read transactions are not cleared, see #2320 and #514.

I'd like to get a better understanding of ledger in JS/WASM to form an opinion on this 🙂

Reply:
Basically, running ledger in JS/WASM can be separated into two stages:
1. Stage 1: Running JS from `ledger.js`, which will load `ledger.wasm` and initialize a WASM runtime for it, analogous to ledger process creation (`preloading` phase in [emscripten-runtime-environment.html#execution-lifecycle](https://emscripten.org/docs/porting/emscripten-runtime-environment.html#execution-lifecycle))
2. Stage 2: Actually running the WASM binary with the runtime, analogous to jumping to the entry point of a loaded process (calling `run()` in [emscripten-runtime-environment.html#execution-lifecycle](https://emscripten.org/docs/porting/emscripten-runtime-environment.html#execution-lifecycle))

One can repeatedly call `run()`/the exported main function (Stage 2) and any static variable will not be reset, similar to:

```c
// js side
void main() {
  while (true) original_main(); // repeatedly call run()
}
// wasm side
#define main original_main
void main () { // ...
```

One may run Stage 1 multiple times when the [`MODULARIZE`](https://emscripten.org/docs/tools_reference/settings_reference.html#modularize)` flag is set. If not, the runtime will initialize only once and might result in issues similar to python ones.